### PR TITLE
Implement lazy indexing, Fix tests

### PR DESCRIPTION
**Pull request by ericcanton**
Fixes indexing, implements a lazy tensor calculation that facilitates faster operations as well as assignment.

All of the unit tests are now passing.

### CODE REVIEW

File: test/nd_list_test.dart [hunk 1]
```diff
@@ -1,9 +1,56 @@
...
+  group('NDIndexResult', () {
+    test('Test resolving', () {
+      final data = [1.0, 2.0, 3.0, 4.0];
+      final ndList = NDList.from<double>(data);
+
+      final result = NDIndexResult.from(ndList);
+      final result2 = result.resolveStep([1, 2, 3], [3]);
+      expect(result2.parentIndices, [1, 2, 3]);
+      expect(result2.parent, equals(ndList));
+      expect(result2.shape, equals([3]));
```
Comment (ftavella) [inline test/nd_list_test.dart:15]: Adding here 
```
expect(result2.evaluate(), equals(NDList.from<double>([2.0, 3.0, 4.0])));
```
will help understand the test better

File: test/nd_list_test.dart [hunk 2]
```diff
@@ -1,9 +1,56 @@
...
+
+      final result3 = result2.resolveStep([1, 2], [2]);
+      expect(result3.parentIndices, equals([2, 3]));
+      expect(result3.parent, equals(ndList));
+      expect(result3.shape, equals([2]));
+      expect(result3.evaluate(), equals(NDList.from<double>([3.0, 4.0])));
+    });
+
+    test('Resolving 3D', () {
```
Comment (ftavella) [inline test/nd_list_test.dart:24]: I like this test but I think it would be nice to check that `evaluate` still works on 3D. Options I'm thinking:
- Change the `ndList` to be something with numbers and then we add an evaluate `expect`
- Do a completely different test where we focus on `evaluate`

File: test/nd_list_test.dart [hunk 3]
```diff
@@ -1,9 +1,56 @@
...
+
+      final subSlice = [1, 5]; // => [5, 17] aka [:, 1, 1]
+      final result3 = result2.resolveStep(subSlice, [2, 1]);
+      expect(result3.parentIndices, equals([5, 17]));
+      expect(result3.parent, equals(ndList));
+      expect(result3.shape, equals([2, 1]));
+    });
+  });
   group('NDList []=', () {
     test('Test can assign 1 element of a 1d NDList', () {
```
Comment (ftavella) [inline test/nd_list_test.dart:81]: It would be nice at the end of this test to check that the final ndList we get from doing all the changes is [5.0, 6.0, 3.0, 7.0]

File: test/nd_list_test.dart [hunk 4]
```diff
@@ -126,7 +173,7 @@ void main() {
       for (var i = 0; i < nRows; i++) {
         for (var j = 0; j < nCols; j++) {
           // NOTE! This is a 1x1 NDList; the element _happens_ to be an NDList with shape [4], but it's not correct to think of ndOfNDs[[i, j]] as the same as it's only element. This is a more complicated example of the Dart difference between 1, [1], and NDList.from<int>([1]).
```
Comment (ftavella) [inline test/nd_list_test.dart]: So we should be careful about the difference between an `ndList of Lists` and an `ndList of ndLists` right? Because I'm thinking if I get a slice of this `ndList of ndLists` I wouldn't get the shape I was expecting.

File: test/nd_list_test.dart [hunk 5]
```diff
@@ -140,7 +187,7 @@ void main() {
     });
   });
 
-  group('NDList<double> indexing', () {
+  group('Basic operators', () {
     test('==', () {
```
Comment (ftavella) [inline test/nd_list_test.dart:344]: Just double checking: is `expect(ndList, equals(ndList));` the same as `expect(ndList == ndList, isTrue);`?

File: lib/nd_list.dart [hunk 6]
```diff
@@ -264,198 +418,158 @@ class NDList<X> {
...
-    if (start < 0) {
-      start += _shape[axis];
-    }
-    if (end < 0) {
-      end += _shape[axis];
-    }
-    if (end < start) {
-      return this._slice(end, start, axis: axis);
-    }
+  static NDIndexResult<Y> _slice<Y>(
```
Comment (ftavella) [inline lib/nd_list.dart:512]: I'm thinking that since this function gave us so many headaches, we should have at least one test per edge case we are trying to get here. That might even help us see if we're missing something. These are the ones I'm thinking:
- `end == start`
- `axis > 0 && priorResult.shape.contains(1)`
- `axis > priorResult.shape.length - 1`
- `start == 0 && end == priorResult.shape[axis]`
- `axis == 0`
- `axis != 0` (and not hitting any of the other ones)

File: lib/nd_list.dart [hunk 7]
```diff
@@ -39,22 +80,132 @@ class NDList<X> {
   final List<int> _shape = [];
 
   List toIteratedList() {
-    return _list.reshape(shape);
+    // Note! Originally from tflite_flutter's ListShape extension.
```
Comment (ftavella) [inline lib/nd_list.dart:81]: I'm having trouble with `toString()` which stems from `toIteratedList()`. When running:
```
import 'package:nd_list/nd_list.dart';

void main() {
  final ndList = NumNDList.zeros([2, 3, 4]);
  ndList.toString();
}
```
I get the following error:
```
Unhandled exception:
Invalid argument(s): Shape and index must have the same length
#0      getLinearIndex (package:nd_list/nd_list.dart:16:5)
#1      NDList._reshape3.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:nd_list/nd_list.dart:146:27)
#2      new _GrowableList.generate (dart:core-patch/growable_array.dart:136:28)
#3      NDList._reshape3.<anonymous closure>.<anonymous closure> (package:nd_list/nd_list.dart:143:21)
#4      new _GrowableList.generate (dart:core-patch/growable_array.dart:136:28)
#5      NDList._reshape3.<anonymous closure> (package:nd_list/nd_list.dart:141:19)
#6      new _GrowableList.generate (dart:core-patch/growable_array.dart:136:28)
#7      NDList._reshape3 (package:nd_list/nd_list.dart:139:45)
#8      NDList.toIteratedList (package:nd_list/nd_list.dart:103:18)
#9      NDList.toString (package:nd_list/nd_list.dart:208:12)
#10     main (file:///C:/Users/tavel/Desktop/Internship-Arcascope/2024_internship/nd_list/test/franco_test.dart:5:10)
#11     _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:297:19)
#12     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

I think it's worth adding tests for this new functionality and all the reshapes

**Review by ftavella — CHANGES_REQUESTED** [review]
Looking good! Thanks for all the comments, it made some things have more sense. I'm suggesting to expand a couple of tests and noted one thing that wasn't working on my end.

File: lib/nd_list.dart (hunk 7 above)
Comment (ericcanton) [reply lib/nd_list.dart:81]: Ah, I expect that would be my attempt to copy what tflite_flutter does to reshape lists. The only reason we had that dependency was for their reshaping. So I probably didn't do that correctly. And literally it's just to print it nicely.

As far as our reshaping, yes we should test that. Especially the -1 functionality. Reshaping our arrays are actually really straightforward.

File: test/nd_list_test.dart (hunk 5 above)
Comment (ericcanton) [reply test/nd_list_test.dart:344]: That's correct, though the preferred method is `expect(ndList, equals(ndList))` since then Dart will explain that the two are not equal, and why, in the test failure. (Versus "This statement is not true." as you get from the other.)

File: test/nd_list_test.dart (hunk 4 above)
Comment (ericcanton) [reply test/nd_list_test.dart]: Yes, these are very different things.

I just checked what happens with `np.ndarray` in python:
```python
import numpy as np

x = np.array([1, 2, 3])
y = np.array([x, 2 * x, 3 * x])
print(y.dtype)  # dtype('int64')
print(y)  # prints 3x3 matrix
```
so indeed NumPy tries to avoid (what would be called) doing something like `y.dtype => np.ndarray`.

I updated the comment here to make it clear this is an `NDList<NDList<double>>` but to be honest, I don't want to change it because I want to make this more predictable via being very literal. Since Dart is strongly typed, I think it makes more sense to not auto type change here.

Is there some change you suggest?

File: lib/nd_list.dart (hunk 7 above)
Comment (ericcanton) [reply lib/nd_list.dart:81]: This should be fixed now, give it a try.

File: test/nd_list_test.dart (hunk 4 above)
Comment (ftavella) [reply test/nd_list_test.dart]: I think it makes sense to keep it as it is!

**Review by ftavella — APPROVED** [review]
Looking great!

File: lib/nd_list.dart (hunk 7 above)
Comment (ftavella) [reply lib/nd_list.dart:81]: Works on my end!
